std/experimental/checkedint.d

```diff
@@ -0,0 +1,1584 @@
+        }
+        else static if (hasMember!(Hook, "onOverflow"))
+        {
+            bool overflow;
+            auto r = opChecked!op(lhs, T(payload), overflow);
+            if (overflow) r = hook.onOverflow!op(42);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else
+        {
+            // Default is built-in behavior
+            auto r = mixin("lhs"~op~"T(payload)");
+            return Checked!(typeof(r), Hook)(r);
+        }
+    }
+
+    // opOpAssign
+    /**
+    */
+    ref Checked opOpAssign(string op, Rhs)(const Rhs rhs)
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        static assert(is(typeof(mixin("payload"~op~"=rhs")) == T));
+
+        static if (hasMember!(Hook, "hookOpOpAssign"))
+        {
+            hook.hookOpOpAssign!op(payload, rhs);
+        }
+        else
+        {
+            alias R = typeof(payload + rhs);
+            auto r = mixin("this"~op~"rhs").payload;
+
+            static if (valueConvertible!(R, T) ||
+                !hasMember!(Hook, "onBadOpOpAssign") ||
+                op.among(">>", ">>>"))
+            {
+                // No need to check these
+                payload = cast(T) r;
+            }
+            else
+            {
+                static if (isUnsigned!T && !isUnsigned!R)
+                    // Example: ushort += int
+                    const bad = r < 0 || r > max.payload;
+                else
+                    // Some narrowing is afoot
+                    static if (R.min < min.payload)
+                        // Example: int += long
+                        const bad = r > max.payload || r < min.payload;
+                    else
+                        // Example: uint += ulong
+                        const bad = r > max.payload;
+                if (bad)
+                    payload = hook.onBadOpOpAssign!op(payload, Rhs(rhs));
+                else
+                    payload = cast(T) r;
+            }
+        }
+        return this;
+    }
+}
+
+// representation
+unittest
+{
+    assert(Checked!(ubyte, void)(ubyte(22)).representation == 22);
+}
+
+/**
+Force all overflows to fail with `assert(0)`.
+*/
+struct Croak
+{
+static:
+    Dst onBadCast(Dst, Src)(Src src)
+    {
+        assert(0, "Bad cast");
+    }
+    Lhs onBadOpOpAssign(string x, Lhs, Rhs)(Lhs, Rhs)
+    {
+        assert(0, "Bad opAssign");
+    }
+    bool onBadOpEquals(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0, "Bad comparison for equality");
+    }
+    bool onBadOpCmp(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0, "Bad comparison for ordering");
+    }
+    typeof(~Lhs()) onOverflow(string op, Lhs)(Lhs lhs)
+    {
+        assert(0);
+    }
+    typeof(Lhs() + Rhs()) onOverflow(string op, Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        assert(0);
+    }
+}
+
+unittest
+{
+    Checked!(int, Croak) x;
+    x = 42;
+    short x1 = cast(short) x;
+    //x += long(int.max);
+}
+
+// ProperCompare
+/**
```

Comment:
When I made this note there were no docs here.

For some reason Github didn't mark the comment as outdated